metadata/src/main/java/org/apache/kafka/controller/ControllerTimeoutDecorator.java

```diff
@@ -0,0 +1,403 @@
+    }
+
+    @Override
+    public CompletableFuture<Void> waitForReadyBrokers(int minBrokers) {
+        return controller.waitForReadyBrokers(minBrokers);
+    }
+
+    @Override
+    public CompletableFuture<Map<ClientQuotaEntity, ApiError>> alterClientQuotas(ControllerRequestContext context, Collection<ClientQuotaAlteration> quotaAlterations, boolean validateOnly) {
+        return controller.alterClientQuotas(context, quotaAlterations, validateOnly);
+    }
+
+    @Override
+    public CompletableFuture<AllocateProducerIdsResponseData> allocateProducerIds(ControllerRequestContext context, AllocateProducerIdsRequestData request) {
+        return controller.allocateProducerIds(context, request);
+    }
+
+    @Override
+    public CompletableFuture<UpdateFeaturesResponseData> updateFeatures(ControllerRequestContext context, UpdateFeaturesRequestData request) {
+        return controller.updateFeatures(context, request);
+    }
+
+    @Override
+    public CompletableFuture<List<CreatePartitionsTopicResult>> createPartitions(ControllerRequestContext context, List<CreatePartitionsTopic> topics, boolean validateOnly) {
+        return controller.createPartitions(context, topics, validateOnly);
+    }
+
+    @Override
+    public CompletableFuture<Void> registerController(ControllerRequestContext context, ControllerRegistrationRequestData request) {
+        return controller.registerController(context, request);
+    }
+
+    @Override
+    public CompletableFuture<AssignReplicasToDirsResponseData> assignReplicasToDirs(ControllerRequestContext context, AssignReplicasToDirsRequestData request) {
+        return controller.assignReplicasToDirs(context, request);
+    }
+
+    @Override
+    public void beginShutdown() {
+        controller.beginShutdown();
+    }
+
+    @Override
+    public int curClaimEpoch() {
+        return controller.curClaimEpoch();
+    }
+
+    @Override
+    public void close() throws InterruptedException {
+        controller.close();
+    }
+
+    @Override
+    public CompletableFuture<List<AclCreateResult>> createAcls(ControllerRequestContext context, List<AclBinding> aclBindings) {
+        return controller.createAcls(context, aclBindings);
+    }
+
+    @Override
+    public CompletableFuture<List<AclDeleteResult>> deleteAcls(ControllerRequestContext context, List<AclBindingFilter> aclBindingFilters) {
+        return controller.deleteAcls(context, aclBindingFilters);
+    }
+
+    // AutoMQ for Kafka inject start
+
+    @Override
+    public CompletableFuture<Integer> getNextNodeId(
+            ControllerRequestContext context,
+            GetNextNodeIdRequestData request
+    ) {
+        return controller.getNextNodeId(context, request);
+    }
+
+    @Override
+    public CompletableFuture<Void> checkS3ObjectsLifecycle(ControllerRequestContext context) {
+        return controller.checkS3ObjectsLifecycle(context).orTimeout(defaultTimeoutInMs, TimeUnit.MILLISECONDS);
```

Comment:
Added timeout only for 2 new methods and that are running by schedule

Reply:
The timeout only complete the CompletableFuture with TimeoutException, the `#checkS3ObjectsLifecycle` is still running in the EventLoop.